mobgap/Reorientation/CorrectSensorOrientationDynamic.py

```diff
@@ -0,0 +1,175 @@
+
+    Parameters
+    ----------
+    data
+        Accelerometer and gyroscope data. The data should be in m/s2 and deg/s.
+    sampling_rate_hz
+        Sampling rate of the data in Hz.
+
+    Returns
+    -------
+    IMU_corrected
+        Corrected IMU data.
+
+    Notes
+    -----
+    Points of deviation from the MATLAB implementation:
+    - The Madgwick filter is applied to the entire data array (walking bout) at once, as opposed to row by row in the
+    original implementation.
+    - The signal is already sliced before calling the present function (CorrectSensorOrientationDynamic).
+    Thus, this script has been simplified by removing a loop and a variable which were not needed.
+    - In MATLAB quaternions are represented as xyzw but python requires wxyz. Thus, the quaternions are adjusted.
+    - Parameters are expressed in the units used in mobgap,specifically, we use m/s^2 instead of g.
+    """
+
+    acc = data.iloc[:, 0:3]
+    gyr = data.iloc[:, 3:6]
+
+    # Madwick is called with the initial orientation specified inside the Matlab Madgwick function used in the Matlab
+    # implementation (UpdateIMUslt2). This choice was made since the first "initial orientation"
+    # specified in the Matlab implementation is overwritten in the UpdateIMUslt2 function.
+    
+    mad = MadgwickAHRS(beta=0.1, initial_orientation=np.array([-0.0863212587840360, -0.639984676042049, 
+                                                               0, 0.763523578361070]))
+
+    chosenacc = acc.copy()
+    chosengyr = gyr.copy()
+
+    # applying the madgwick algo to the entire data array
+    data = pd.concat((chosenacc, chosengyr), axis=1)
+    mad = mad.estimate(data, sampling_rate_hz=sampling_rate_hz)
+    quaternion = mad.orientation_.iloc[1:, :]
+    quaternion = quaternion.to_numpy()
+
+    # Adjust quaternion from x, y, z, w to w, x, y, z
+    quaternion = quaternion[:, [3, 0, 1, 2]]
+
+    av = acceleration(data, quaternion)
+
+    # Principal component analysis
+    pca_acc = PCA()
+    pca_acc.fit(chosenacc)
+    pca_coef = pca_acc.components_
+    newacc = chosenacc.values @ pca_coef.T
+
+    pca_gyr = PCA()
+    pca_gyr.fit(chosengyr)
+    pca_coef_gyr = pca_gyr.components_
+    newgyr = chosengyr.values @ pca_coef_gyr.T
+
+    if np.mean(newacc[:, 0]) < 0:
+        newacc[:, 0] = -newacc[:, 0]
+
+    # av_magpca
+    av_magpca = av.copy()
+    av_magpca[:, 0] = newacc[:, 0]
+
+    # gyr_magpca
+    gyr_magpca = chosengyr.to_numpy().copy()
+    gyr_magpca[:, 0] = newgyr[:, 0]        # Yaw
+
+    # Standardization
+    sig1 = (av[:, 2] - np.mean(av[:, 2])) / np.std(av[:, 2])
+    sig2 = (newacc[:, 1] - np.mean(newacc[:, 1])) / np.std(newacc[:, 1])
+    sig3 = (newacc[:, 2] - np.mean(newacc[:, 2])) / np.std(newacc[:, 2])
+    sig4 = (av[:, 1] - np.mean(av[:, 1])) / np.std(av[:, 1])
+
+    # Assigning av_pca and gyr_pca
+    av_pca = av_magpca.copy()
+    gyr_pca = gyr_magpca.copy()
+
+    # Calculating correlations
```

Comment:
Can you add some comments on what is happening here? I am confused

Reply:
I added comments in the whole code below this point